### ISSUE TITLE
Request for feature (?),  a way to sample from large jsonl manifests.

### ISSUE THREAD

**Issue by danpovey** [lhotse-speech/lhotse#1096]
@pzelasko for some recent language modeling work I have been using a very simple type of data loader and I find it very refreshing to work with.  The data loader is completely stateless: it makes no attempt to sync any kind of state, but just draws completely randomly from the full data each time.   If someone resumes training, we just use the start-batch index as part of the random seed, so it won't repeat the already seen data; but it also doesn't make an attempt to see the exact same data that you would have seen if you had iterated from the start.  [There are ways you could fix this fairly easily, though.]

In my case, for language modeling the way it works (https://github.com/k2-fsa/icefall/pull/1141/files#diff-d641dd2a3ec42b15b0f36fd57011a184ad8119a1fa478d5e4f1485b47113ab73) is we write a kind of manifest containing the list of training files and their length in bytes; and all the data loaders read this; they use a nn.random.multinomial to randomly select the training file with prob proportional to its length; then choose a random chunk within the file.  I decided that it was too much of a hassle to deal with epochs, so there is no concept of epoch.  In training mode, it just loops infinitely and the training loop decides when to exit, based on the total amount of data that has been seen.   We make the Eden scheduler depend on the number of tokens seen in this case, not on the epochs; this can be done from train.py without changing the optimizer code.

In eval mode, it just loops enough to see the data about once; but in reality you'd use a different data loader for validation data.

The way I propose to do this for not-compressed jsonl manifests, is as follows.

Firstly, the user would provide to this data loader either a list of jsonl files, or a list of (jsonl, scale) where `scale` is a real-valued factor that you can use to upweight or downweight the contribution of that jsonl file.  Each time we sample, we sample the jsonl file with probability (number_of_lines_in_that_jsonl_file * scale), and then randomly choose the line number within the jsonl file.

The data loader needs to know, for each jsonl file, the list of start-positions, in bytes, of the lines in that file; so it can sample the line index and then seek to that position.  We could cache this in a binary .jsonl.idx file for each of the .jsonl files (or just back off to having each data loader compute this list at the start, e.g. in case the file system is not writable; this won't be too slow.)  It can include the length in bytes of the jsonl file, for purpose of invalidating old cached data.  If the size of the jsonl.idx file is above some cutoff and it's on disk, the data loader could just store the length of the list, keep a file handle open, and fseek to the appropriate position in the jsonl.idx file each time it needs a line; the system can take care of caching.  We don't need to worry about a little extra latency because the data loader workers now require relatively little memory.

We can view this process as a random source of lines for a manifest cached in memory; I suppose we'd use the dynamic batching on top of this.  IMO the advantages of this are:

  - Makes the code much simpler, and easier to analyze; no need to worry about seeing data just once.
  - Less I/O for reading manifests since we don't need to duplicate things and then discard 3/4 or 7/8 of the manifests.
  - Potentially much less memory used in data loaders
  - Less hassles when you restart in the middle of training; can simplify training code.
  - Should be easier to mix and match data and adjust the probabilities
  - Easy for users to understand what is going on.
  - Makes training code simpler since we abandon the concept of an epoch.

I think the reduction in complexity will be important for the long-term viability of what we are doing, because the way it is done now is hard enough to understand that it might be hard for others to maintain the code in future.

**Comment by pzelasko** [comment]
Sorry for the late answer, I've been traveling and just got back recently. I should have a bit more time in my calendar soon, so I'll re-read carefully and see how to get something like this added.

**Comment by pzelasko** [comment]
Please check out the PR https://github.com/lhotse-speech/lhotse/pull/1102
